### PR TITLE
fix(catalog): recognize YYMMDD/MMDD/YYMM dated model id suffixes (#3024)

### PR DESCRIPTION
**Pull request by kaicot**
Fixes the safely-fixable half of #3024, in the scope you picked (widen the suffix format, keep the fold one-directional) with the `YY = 2\d` tightening from your review.

## Summary

`isDatedVariantId` only accepted an 8-digit `YYYYMMDD` suffix, so the dated-alias fold in `mergeConfiguredModelsIntoLiveCatalog` never fired for providers that publish shorter release dates. Enumerating every numeric-suffixed id pair on a real multi-provider install, the `\d{8}` rule matched **0 of 26** — so DeepSeek, Kimi, Mistral, Qwen and Solar aliases all fell through to `droppedConfiguredIds` even when a live row for the same model was present.

This widens the suffix to the formats upstreams actually publish, behind a calendar guard:

| Format | Example | Before | After |
|---|---|---|---|
| `YYYYMMDD` | `claude-haiku-4-5-20251001` | yes | yes |
| `YYMMDD` | `solar-pro4-260806`, `syn-pro-251021` | no | yes |
| `MMDD` | `deepseek-v4-pro-0813`, `kimi-k2-0905`, `gpt-3.5-turbo-0613` | no | yes |
| `YYMM` | `mistral-large-2407`, `qwen3-235b-a22b-2507` | no | yes |
| `YYYY-MM-DD` | `gpt-4o-2024-08-06` | no | no (out of scope) |
| `MM-DD` | `gemini-2.5-pro-preview-05-06` | no | no (out of scope) |

The matcher now parses and validates real calendar dates. `YYYYMMDD` and `YYMMDD` apply Gregorian leap-year rules, `MMDD` is accepted only when the date occurs in at least one calendar year, and `YYMM` remains restricted to the 2020s. Impossible values such as `20250229`, `250229`, and `0431` are rejected. Ordinary numeric suffixes `-2048`, `-4096`, and `-8192` remain rejected. `-1024` is still the one irreducible collision because October 24th is a real date; an explicit regression test documents that accepted cost.

Two things this deliberately does **not** do:

- **The fold stays one-directional** (`configured=base` → `live=dated`). A configured id the provider no longer lists must not be retained on the strength of a format match alone — the fold has no callability signal, and #1690 is the right place for an explicit operator allow-list. This means `alibaba-token-plan-intl/deepseek-v4-pro-0813`, the id that opened #3024, **is still dropped**. That is the deliberate call from your review, not an oversight.
- **No change to `shouldRetainConfiguredProviderModel` or the `retainModels` path**, so this does not touch or conflict with the #1690 train (#2860 / #2122).

The 8-digit path is intentionally stricter than the historical bare `\d{8}` rule: impossible calendar dates no longer fold. This follows the review decision and is covered by month-end and leap-day regression tests.

## Verification

Revalidated after rebasing onto the then-current `dev` commit `44b4de39d` on Windows 11 with Bun 1.4.0:

```
bun test tests/codex-catalog.test.ts  # 251 pass, 0 fail
bun run typecheck                      # exit 0
bun run privacy:scan                   # Privacy scan passed
git diff --check upstream/dev...HEAD   # exit 0
```

The CodeRabbit finding is covered by explicit valid/invalid leap-day and month-end cases for `YYYYMMDD`, `YYMMDD`, and `MMDD`.

The repository-wide `bun run test` continued making progress but reached the runner's 900-second Windows limit. Failures observed outside this catalog change were Windows ACL/temporary-directory locking and existing 5-second integration-test timeouts. As a control, the Kiro file that failed with `EBUSY` in the parallel suite passed alone (15 pass, 0 fail). `responses-state.test.ts` still fails alone when its ACL-backed spill path is unavailable on this D: volume. I am therefore leaving the local-all-green readiness box unticked and treating GitHub CI as the authority, as requested in review.

Tests remain table-driven in the existing `isDatedVariantId` section. Added calendar regressions include valid `20240229` / `240229` / `0229`, invalid `20250229` / `250229`, and invalid April 31 values in all applicable widths.

## Checklist
- [x] Scope stays focused and avoids unrelated cleanup.
- [x] Docs or release notes were updated when needed. (No user-facing docs describe this fold; the rationale lives in the JSDoc on the pattern.)
- [x] Security-sensitive changes were reviewed for secrets, auth, and unsafe defaults. (Pure predicate change; no auth, network, or config surface touched. `bun run privacy:scan` passes.)

## Review readiness checklist

This PR stays in draft until every box below is ticked. Tick all four boxes once the requirements are met:

- [ ] All CI tests are green on my local testing.
- [x] I pushed my PR to the latest dev commit.
- [x] I resolved all correct Codex and CodeRabbit findings.

- [ ] My PR is ready for review.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved recognition of dated model aliases using multiple date formats.
  * Prevented valid aliases from being incorrectly excluded from configured model listings.
  * Continued rejecting version numbers, context sizes, invalid dates, and unrelated suffixes.

* **Tests**
  * Expanded coverage for supported date formats and edge cases, including the accepted `-1024` suffix behavior.

### CODE REVIEW

**Comment by lidge-jun** [comment]
## 리뷰 · 우선순위 66 / 80

설명

이 PR은 열린 이슈 #3024 의 「안전하게 고칠 수 있는 절반」만 다룬다. 지금 `dev` (`7666f7d2a`) 의 `src/codex/catalog/provider-fetch.ts` 안 `isDatedVariantId` 는 아직 `/^\d{8}$/` 한 줄이다. 설정에 짧은 이름(base)이 있고 라이브 카탈로그에 날짜 붙은 이름(dated)이 있을 때만 그 둘을 같은 모델로 접어 주는 함수다. 여덟 자리 `YYYYMMDD` 만 날짜로 인정하니, 실제로 DeepSeek·Kimi·Mistral·Qwen·Solar 가 쓰는 `YYMMDD` / `MMDD` / `YYMM` 접미사는 전부 날짜가 아니라고 판정되고, 설정 쪽 짧은 id 가 `droppedConfiguredIds` 로 떨어진다. 작성자가 실제 멀티 프로바이더 설치에서 숫자 접미사 쌍 26개를 세어 보니 옛 규칙에 맞은 것이 0개였다고 한다.

고치는 방식은 단순하다. 월·일 모양 가드가 있는 정규식 네 갈래(`YYYYMMDD` / `YYMMDD` / `MMDD` / `YYMM`) 로 넓힌다. 연도 두 자리는 `2\d` 로 묶어서 `-1301` 같은 불가능한 YYMM 을 막고, `-2048` / `-4096` / `-8192` 같은 컨텍스트 크기 접미사는 월이 20·40·81 이라서 자연히 거절된다. 반대로 `-1024` 는 10월 24일로 읽혀서 통과한다. 이건 버그가 아니라 MMDD 를 허용하는 한 피할 수 없는 충돌이고, 테스트로 그 비용을 고정해 두었다. 하이픈이 들어간 ISO 날짜(`-2024-08-06`, `-05-06`) 는 일부러 범위 밖이다.

더 중요한 설계 선택은 「접기는 한쪽 방향만」이다. 설정=짧은이름 → 라이브=날짜이름 만 접고, 설정=날짜이름 → 라이브=짧은이름 은 접지 않는다. 그래서 이슈를 연 예시 `alibaba-token-plan-intl/deepseek-v4-pro-0813` 처럼 설정에 날짜 id 만 남아 있고 프로바이더가 짧은 id 만 주는 경우는 이 PR 후에도 여전히 떨어진다. 그건 실수가 아니라 #1690 retain/allow-list 쪽 일감으로 남기라는 메인테이너 리뷰를 따른 것이다. `shouldRetainConfiguredProviderModel` 과 retainModels 경로도 안 건드린다. 테스트는 `tests/codex-catalog.test.ts` 에 표 형태로 포맷별 통과·거절·컨텍스트 크기·1024 충돌·역방향 거절을 추가한다. types.ts/config.ts 분할과 무관하다.

점수 66/80. 범위가 분명하고 HEAD 의 실제 병목(여덟 자리만 인정)을 정확히 겨냥하며, 알려진 비용을 테스트로 박아 두었다. 다만 이슈 제목의 「callable 인데 떨어짐」 전체 중 설정=dated 쪽은 일부러 안 고쳤고, 체크리스트 네 칸이 아직 비어 있으며, 작성자도 전체 prepush 를 끝까지 못 돌렸다. CI 초록이 권위다.

라인 provider-fetch.ts ~938 (HEAD) - 현재는 `/^\d{8}$/` 한 줄. PR 이 여기만 교체한다. 범위 좋다.
경로 DATED_VARIANT_DD = 3[01] - 2월 31일 같은 「달력에 없는 날」도 모양만 맞으면 통과한다. 진짜 달력 검증은 아니다. 의도가 「월·일 자리 모양」이면 문서/주석에 그 한계를 한 줄 더 적어도 좋다.
경로 8자리 갈래 - 이제 달력 모양을 요구하므로 예전의 「아무 여덟 숫자」보다 조금 더 빡세다. 작성자가 PR 본문에서 이미 물어봤다. `-99999999` 같은 가짜 여덟 자리는 더 이상 접히지 않는다.
경로 역방향 접기 없음 - `deepseek-v4-pro-0813` 설정 vs `deepseek-v4-pro` 라이브는 계속 false. 이슈 #3024 제목 사례 중 설정=dated 쪽은 고치지 않는다. 의도된 잔여다.
PR 본문 체크리스트 - CI 로컬 완료 / 최신 dev 기준 push / Codex·CodeRabbit 반영 / ready 네 칸이 비어 있다. draft 해제 전에 채워야 하는 저장소 관례다.
검증 주장 - typecheck 와 catalog 관련 테스트·privacy:scan 통과는 적혀 있으나 전체 `test:changed` / prepush 는 작성자가 중단했다. CI 결과를 본다.

메인테이너의 판단이 필요한 지점

- 8자리 경로를 달력 가드로 조이는 것(옛 동작 대비 약간 타이트)을 받아들일지, 아니면 8자리만 예전처럼 `\d{8}` 로 남겨 둘지.
- `-1024` 가 MMDD 로 통과하는 비용을 그대로 둘지. 작성자·테스트는 「알려진 비용」으로 고정했다.
- 설정=dated / 라이브=base 잔여(#3024 원래 예시)를 #1690 으로 넘기는 결정을 이번에도 유지할지. 유지하면 이 PR merge 후에도 #3024 를 바로 닫지 말고, 「접미사 확장 half landed, reverse-fold/retain 은 #1690」이라고 이슈에 남겨야 한다.
- 체크리스트가 비어 있는 동안 ready-for-review 로 볼지, CI 초록만으로 충분하다고 볼지.

너의 추천

CI 가 초록이면 merge 후보로 둔다. 범위 확대해서 역방향 접기나 retainModels 를 이 PR에 싣지 말 것 — #1690 과 섞이면 분할 캠페인·다른 카탈로그 PR 과 충돌한다. merge 후 #3024 에는 「suffix half landed via #<이 PR>, reverse/retain 잔여」 댓글을 남기고 이슈는 열어 두거나 부분 완료로 표시한다. 8자리 조임과 -1024 비용은 작성자 제안대로 받아들이는 쪽이 일관적이다.

이 댓글은 grok-bot이 작성했습니다

File: src/codex/catalog/provider-fetch.ts [hunk 1]
```diff
@@ -935,9 +935,36 @@ export function resolveComboCatalogMember(
   };
 }
 
+const DATED_VARIANT_MM = String.raw`(?:0[1-9]|1[0-2])`;
+const DATED_VARIANT_DD = String.raw`(?:0[1-9]|[12]\d|3[01])`;
```
[the remark this answers is not in the document]
Comment (kaicot) [reply src/codex/catalog/provider-fetch.ts]: Fixed in `0a6393d20`. The suffix matcher now parses calendar components instead of relying on shape-only day ranges: `YYYYMMDD` and `YYMMDD` apply Gregorian leap-year rules, while `MMDD` is accepted only when the date occurs in at least one year. Regression cases cover valid leap days plus invalid non-leap February 29 and April 31 values. The documented `-1024` collision remains accepted. Revalidated: `tests/codex-catalog.test.ts` 251 pass / 0 fail, typecheck and privacy scan pass.

**Comment by lidge-jun** [comment]
Superseded by #3100, which cherry-picks this commit with your authorship intact and adds three merge-loop regressions carried from #3041.

Your calendar matcher is the better of the two: it rejects `20250229` and `0231`, which #3041's bare `/^\d{8}$/` plus `day <= 31` both accept. #3100 keeps it unchanged.

What it adds is where the direction guard lives. Your `does not fold configured=dated against live=base` asserts on the predicate, and that assertion is true of every implementation — including one whose merge loop calls the predicate a second time with the arguments swapped, which is what #3041 first proposed. Mutation-checked: adding `|| isDatedVariantId(candidate.id, live.id)` to the merge loop leaves your predicate test green and fails only the new merge-loop test.

Please keep an eye on #3100; if you would rather land this yourself, say so and I will close mine instead.

Triaged in the 2026-08-31 non-priority-70 bug round.

**Comment by lidge-jun** [comment]
Landed via maintainer replacement #3100 at b46164e7858e35c50f8c1a0605d4d333fd434862. The calendar matcher from this PR and the directional merge-loop guards from #3041 were combined on current dev with exact-head CI. Closing this superseded source PR.
